{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "from Bio import AlignIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Program versions:\n",
    "- `clustalo` 1.2.4\n",
    "- `clustalw` 2.1\n",
    "- `mafft` v7.520\n",
    "- `muscle` 5.1.linux64\n",
    "- `kalign` 2.04\n",
    "- `t_coffee` Version_12.00.7fb08c2\n",
    "- `prank` v.170427\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Run 6 algos (clustalw, muscle, mafft, kalign, t_coffee, prank) on `.fasta` file with 10 sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.68 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "6.06 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "2.53 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "343 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "24.8 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "1min 16s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "time_dct = {}\n",
    "time_dct[\"clustalw\"] = %timeit -n1 -r1 -o !clustalw -INFILE=SUP35_10seqs.fa -OUTPUT=FASTA -OUTFILE=02_SUP35_10seqs_clustalw.fa > logs/logs_clustalw.txt\n",
    "time_dct[\"muscle\"] = %timeit -n1 -r1 -o !muscle -align SUP35_10seqs.fa -output 02_SUP35_10seqs_muscle.fa 2> logs/logs_muscle_std2.txt 1> logs/logs_muscle_std1.txt\n",
    "time_dct[\"mafft\"] = %timeit -n1 -r1 -o !mafft --auto SUP35_10seqs.fa >02_SUP35_10seqs_mafft.fa 1> logs/logs_mafft_std1.txt 2> logs/logs_mafft_std2.txt\n",
    "time_dct[\"kalign\"] = %timeit -n1 -r1 -o !kalign <SUP35_10seqs.fa >02_SUP35_10seqs_kalign.fa 2> logs/logs_kalign.txt\n",
    "time_dct[\"tcoffee\"] = %timeit -n1 -r1 -o !t_coffee -infile=SUP35_10seqs.fa -outfile=02_SUP35_10seqs_tcoffee.fa -output=fasta_aln 2> logs/logs_t_coffee_std2.txt 1> logs/logs_t_coffee_std1.txt\n",
    "time_dct[\"prank\"] = %timeit -n1 -r1 -o !prank -d=SUP35_10seqs.fa -o=02_SUP35_10seqs_prank.fa -codon > logs/logs_prank.txt # codon for codon alignment\n",
    "\n",
    "for k in time_dct:\n",
    "    time_dct[k] = round(time_dct[k].average, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'clustalw': 2.68,\n",
       " 'muscle': 6.06,\n",
       " 'mafft': 2.53,\n",
       " 'kalign': 0.34,\n",
       " 'tcoffee': 24.78,\n",
       " 'prank': 76.88}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_dct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Get comparative table for every aln algo with time and aln length"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Simple func to get aln length from fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_aln_len(path: str) -> int:\n",
    "    with open(path) as f:\n",
    "        aln_len = 0\n",
    "        for line in f:\n",
    "            if line.strip().startswith(\">\") and aln_len != 0:\n",
    "                break\n",
    "            else:\n",
    "                aln_len += len(line.strip())\n",
    "    return aln_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "length_dct = {}\n",
    "\n",
    "for file in os.listdir():\n",
    "    if file.startswith(\"02_SUP35_10seqs\"):\n",
    "        length_dct[file.split(\"_\")[-1].split(\".\")[0]] = get_aln_len(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'clustalw': 2177,\n",
       " 'muscle': 2368,\n",
       " 'tcoffee': 2229,\n",
       " 'prank': 2453,\n",
       " 'mafft': 2185,\n",
       " 'kalign': 2174}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "length_dct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>time</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>kalign</th>\n",
       "      <td>0.34</td>\n",
       "      <td>2174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mafft</th>\n",
       "      <td>2.53</td>\n",
       "      <td>2185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clustalw</th>\n",
       "      <td>2.68</td>\n",
       "      <td>2177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>muscle</th>\n",
       "      <td>6.06</td>\n",
       "      <td>2368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tcoffee</th>\n",
       "      <td>24.78</td>\n",
       "      <td>2229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prank</th>\n",
       "      <td>76.88</td>\n",
       "      <td>2453</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           time  length\n",
       "kalign     0.34    2174\n",
       "mafft      2.53    2185\n",
       "clustalw   2.68    2177\n",
       "muscle     6.06    2368\n",
       "tcoffee   24.78    2229\n",
       "prank     76.88    2453"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame({\"time\": time_dct, \"length\": length_dct}).sort_values(by=[\"time\", \"length\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In terms of time the best algo is kalign, as well, as in aln length. Smaller length usually means less gaps, but one should consider biological sense behind any alignment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Let's tale a look at `SUP35_10seqs_strange_aln.fa`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">SUP35_Kla_AB039749\n",
      "ATGTCAG---------------------------ACCAACAAAATCAAGACCAAGGGCAAGGCCAAGGTTACAATCAGTATAACCAATATGGCCAGTACAACCAG----TACTACAACCAACAGGGCTATCAAGGCTACAACGGCCAACAAGGTGCTCCTCAAGGCTACCAAGCATATCAAGCTTATGGACAGCAGCCTCAAGG---------------------AGCCTACCAGGGCTAC---AACCCTC--AACAAGCTCAAG-----GCT---ACCAACCTTACCAGGGCT---ACAATGCTCAGCAA------CAAGGTTACAAC-----------GCTCAG------------------------CAAGGCGGTCA-CAACAATAACTACAATAAAAA--------TTATAATAATAA---AAACAGTTACAATAACTATAATAAGCA-GGGTTATCAAGGTGCTCAAGGATATAATGCACAACAGCCAACCGGTTACGCTGCTCCAGCACAGTCTTCATCCCAGGGTATGACT-----TTGAAAGATTTCCAAAA-------CCAAC-AAGGCAGTACTAATGCAGCCAAGCCAAAGCCTAAGTTAAAGTTGGCCTCTAGCTCTGGTATTAAGTTAGTAGGTG---CCAAGAAACCTGTAGCA------CCCAAAACTGAGAAAA-----------------CTGATGAATC--CAAGGAAGCAACTAAA------------------------------------------------ACTACCGACGACAACGAAGAAGCACAATCTGAATTGCCCAAAATTGATG-ATTTGAAAATCTCAGAGGCTGAAAAACCAAAAACTAAGGAGAATACCCCATCTGCTGATGATACTTCCTCAGAGAAGACTACCAGCGCTAAGGCAGACACATCTACAGGAGGAGCGAACTCCGTGGAT------GCTCTAATCAAGGA---ACAAGAAGATGAGGTTGACGAAGAAGTCGTTAAAGATATGTTTGGTGGTAAGGATCATGTTTCCATCATTTTCATGGGTCACGTCGATGCT--GGTAAGTCAACAATGGGTGGTAACTTGTTATATCT--GAC----------------------TGGTTCTGTGGATAAAAGAAC-CGTTGAGAAATATGAGAGAGAGGCTAAAGAGGCTGGTAGACAAGGTTGGTATTTATCATGGGTGATGGATACCAACAAAGAAGAAAGA--AACGACGGTAAAACCATTG---AAGTGGGTAGAGCGTACTTTGAAACTGAAAAGAGACGTTACACTATTTTGGATGCTCCCGGTCACAAAATGTATGTTTCTGAAATGATTGGTGGTGCATCTCAAGCCGATATTGGTATTTTGGTTATTTCTGCTAGAAAGGGTGAATATGAAACTGGTTTCGAAAAGGGTGGTCAAACTCGTGAGCACGCTT---TATTAGCCAAGACACAAGGTGTCAACAAAATGATTGTAGTTATCAACAAGATGGATGAT--CCAACTGTGGGATGGGATAAGGAAAGATATGATCACTG----TGTTGGTAACTTGACAAACTTTTTGAAGGCTGTGGGTTACAATGTTAAGGAGGACGTCATTTTCATG----------------CCAGTGTCTGGTTACACAGGTGCAGGTTTGAAGGAACGTGTCGATCCTAAGGACTGTCCATGGTATACTGGTCCATCTTTATTAGAATATCTTGACAATATGAAGACTACTGATCGTCATATCAATGCTCCATTCATGCTTCCAATTGCTTCTAA--GAT-----GAAGGACATGGGTACTGTTGTGGAAGGTAAAATCG--------AATCTGGACACATTAGAAAGGGTAACCAAACTTTACTAATGCCTAACAGGACCTCTGTTGAAATTCTGACCA--TTTATAACGAAACTGAAAGCGAAGTTGACA--TGGCTGTTTGTGGTGAGCAGGTTAG----ATTAAGAATTAAAGGTGTCGAAGAAGAAGAAATTTCTGCTGGTTTCGTTCTAACCTCTCCAAAAAACCCA-----GTTAAGAATGTAACGAGATTTGTGGCTCAAATTGCTA--------TTGTCGAATTGAA--ATCGATCATGTCTGCTGGTTTCTCGTGCGTTATGCATATTCATACAGCTATCGAAGAAGTCACCGTCACAAGATTGCTTCACAAGCTTGAGAAGGGGTCA-------AACAGAAAATCAAAGAAGCCTCCAGCATTTGCTAAAAAGGGTATGAAGATCATTGCCGTTATCG--AGACTAATGAACCGGTATGTGTTGAAACATACGAT---GATTACCCACAATTGGGTAGATTCACTTTAAGAGATCAAGGTACCACCATTGCTATCGGTAA-------GATTGTGAAAATCCTTGAAAATTGA---\n",
      ">SUP35_Agos_ATCC_10895_NM_211584\n",
      "ATGTCGGAGGAAGATCAAATTCAATCGCAAGGCAACGACCAAGGCCAGTCGCAAGCCAAGGATCAAGGTCAAAATCAAGGTCAGGGGCAGCAAAATTTCGGCCAA----TACTACAACCCAAGTAACTTCCAGAATTACCAGGGGTA------TGTGCCTCAGGG------AGGTTACCAGGCGTATGGC---CAACAGGCAGGTGGGTACCAAGGCTATGCGCAATACAACCAGCAAGCCGGCGGGTACC--AGGGTTATCAGG-----GAT---ACCAGCAGTACAACCCAGCACAAGCTGGCTACCAAGGTTACCAGCAATACAAT-----------GCGCAAGGTGGGTAT---------------CAGAGCTACAAGCAGTACAACTCACAGCCACAGGGGAACCGTAAGGGGAACCA---GAGCTATGGTTACGGACA----------GGGCCAGTCAGCCACCG------------------------------------------------CT------CCGGTGACG--------CTGAACAACTTTGAGAAGGGAACTGTGCCGAATGCGA-----CTGCTCCAAAACCAAAGAAGACCCTCAAGTTGGCTTCCAGCTCCGGTATCAAGCTAGTTGGTG---CTAAGAAACCTGTCGCG------------------AAGAAAGAAGA--GGCTAAGGCGGAGGAGCCAACAAAGGAGGAGCCTAA-----GAGCAGCGCAGAGGGCGCTCCTAAGAGTGAGGATGCTACCGCTTCCTCTGAAGATAAGGCAG---------------TCCCAAGTATTGAGA-AGCTAAGTATTTCGGAAGC------------------------------CGATACCGCGAAGAAAGACACAGCGGACGCGGCCGGTGCC---------------------------ACCTCATCGGAT------GCTCTGATCAAGGA---ACAGGAAGACGAAGTTGATGAAGAAGTTGTGAAGGACATGTTTGGTGGAAAGGACCATGTTTCTATCATCTTTATGGGTCACGTCGATGCG--GGGAAATCCACCATGGGCGGAAACTTGTTGTATCT--AAC----------------------CGGCTCGGTAGACAAAAGAAC-AGTTGAGAAGTATGAGAAGGAGGCGAAGGAAGCAGGGAGACAGGGTTGGTACTTGTCTTGGATTATGGATACAAATAAGGAGGAGAGA--AACGATGGTAAGACCATTG---AAGTGGGTAGATCCTACTTTGAAACTGAAAAGAGACGGTATACCATTTTGGATGCACCAGGCCACAAAATGTACGTCTCTGAAATGATTGGTGGTGCGTCTCAAGCCGATGTTGGTATATTGGTCATCTCAGCCAGAAAGGGTGAGTACGAGACTGGTTTTGAGAGAGGTGGACAAACGCGTGAGCATGCGC---TATTAGCGAAGACCCAGGGTGTCAATAAAATGGTGGTCGTCGTGAACAAAATGGATGAC--CCCACGGTAAACTGGGACAAGGCACGTTATGACCAGTG----TATTAAGAACGTCTCCAATTTCTTGCAAGCTATTGGCTACAACGTTAAGGAGGATGTTATGTATATG----------------CCTGTTTCTGGGTTTACTGGTGCTGGTCTGAAGGACCGCGTTGACAAGAAGGACTGTCCTTGGTACGATGGTCCTTCCCTTTTGGAATACCTAGACAACATGCAGCATGTTGACCGCTTCATCAACGCGCCTTTCATGCTTCCAATTGCAAGCAA--GAT-----GAAGGATATGGGTACAGTGGTCGAAGGTAAAATCG--------AATCTGGTCATATTAAGAAGGGTAATCAAACACTACTGATGCCAAACAAAATTCCAGTGGAGATTCTAGCCA--TCCAAAATGAGACCGAACAGGAAGTTGATA--TGGCTGTGTGTGGTGAGCAGGTCAG----ATTGAGACTGAAGGGAGTTGAGGAGGAAGACATTTCTGCAGGTTTTGTTTTGACTTCGCCAAAGAACCCT-----GTCAAGAACGTCACGAAGTTTGTTGCCCAGATTGCCA--------TTGTGGAATTGAA--GTCCATTATGTCAGCAGGTTTCTCCTGTGTGATGCACGTGCATACCGCTATTGAGGAAGTCTCTATCACTAGACTACTTCATAAGCTGGAGAAGGGCACC-------AACAGAAAGTCCAAGAAGCCACCTGCCTTTGCTAAGAAGGGCATGAAGATAATAGCGGTCCTTG--AGACAGAGGAGCCAGTATGTGTGGAAACCTACCAG---GATTACCCACACTTGGGTAGATTCACATTGAGAGACCAGGGTATTACAATCGCCATCGGTAA-------GATTGTGAAGATCTTGGAA---TGA---\n",
      ">SUP35_Scer_74-D694_GCA_001578265.1\n",
      "ATGTCGG---------------------------ATT--CAAA-------------------CCAAGGCAACAATCAGCAAAAC---TACCAGCAATACAGCCAG----AACGGTAACCAACAA------CAAGGTAACAACAGATA------------CCAAGGTTATCAAGCTTACAATGCTCAAGCC---CAACCTGCAGGTGGGTACTAC---------CAAAATTACCAAGGTTATTCTGGGTACC--AACAAGGT---G-----GCT---ATCAACAGTACAATCCCG---ACGCCGGTTACCAG------CAACAGTATAAT-----------CCTCAAGGAGGCTATCAACAGTACAATCCTCAAGGCGGTTATCAGCAGCAATTCAATCCACAAGGTGGCCGTGGAAATTACAA---AAACTTCAACTACAATAACAATTTGCA-AGGATATCAAGCTGGTT------------------------------------TCCAACCACAGTCT------CAAGGTATGTCT-----TTGAACGACTTTCAAAA-------GCAACAAAAGCAG----GCCGCTCCCAAACCAAAGAAGACTTTGAAGCTTGTCTCCAGTTCCGGTATCAAGTTGGCCAATGCTACCAAGAAGGTTGGCACAAAACCTGCCGAATCTGATAAGAAAGAGGA--AGAGAAGTCTGCTGAAAC--CAAAGAACCAACTAAAGAGCCAACAAAGGTCGAAGAACCAGTTAAAAAGGAGGAGAAACCAGTCCAGACTGAAGAAAAGACGGAGGAAAAATCGGAACTTCCAAAGGTAGAAG-ACCTTAAAATCTCTGA---------------------------------ATCAACACATAATACC---------AACAATGCCAATGTT---------------------------ACCAGTGCTGAT------GCCTTGATCAAGGA---ACAGGAAGAAGAAGTGGATGACGAAGTTGTTAACGATATGTTTGGTGGTAAAGATCACGTTTCTTTAATTTTCATGGGTCATGTTGATGCC--GGTAAATCTACTATGGGTGGTAATCTACTATACTT--GAC----------------------TGGCTCTGTGGATAAGAGAAC-TATTGAGAAATATGAAAGAGAAGCCAAGGATGCAGGCAGACAAGGTTGGTACTTGTCATGGGTCATGGATACCAACAAAGAAGAAAGA--AATGATGGTAAGACTATCG---AAGTTGGTAAGGCCTACTTTGAAACTGAAAAAAGGCGTTATACCATATTGGATGCTCCTGGTCATAAAATGTACGTTTCCGAGATGATCGGTGGTGCTTCTCAAGCTGATGTTGGTGTTTTGGTCATTTCCGCCAGAAAGGGTGAGTACGAAACCGGTTTTGAGAGAGGTGGTCAAACTCGTGAACACGCCC---TATTGGCCAAGACCCAAGGTGTTAATAAGATGGTTGTCGTCGTAAATAAGATGGATGAC--CCAACCGTTAACTGGTCTAAGGAACGTTACGACCAATG----TGTGAGTAATGTCAGCAATTTCTTGAGAGCAATTGGTTACAACATTAAGACAGACGTTGTATTTATG----------------CCAGTATCCGGCTACAGTGGTGCAAATTTGAAAGATCACGTAGATCCAAAAGAATGCCCATGGTACACCGGCCCAACTCTGTTAGAATATCTGGATACAATGAACCACGTCGACCGTCACATCAATGCTCCATTCATGTTGCCTATTGCCGCTAA--GAT-----GAAGGATCTAGGTACCATCGTTGAAGGTAAAATTG--------AATCCGGTCATATCAAAAAGGGTCAATCCACCCTACTGATGCCTAACAAAACCGCTGTGGAAATTCAAAATA--TTTACAACGAAACTGAAAATGAAGTTGATA--TGGCTATGTGTGGTGAGCAAGTTAA----ACTAAGAATCAAAGGTGTTGAAGAAGAAGACATTTCACCAGGTTTTGTACTAACATCGCCAAAGAACCCT-----ATCAAGAGTGTTACCAAGTTTGTAGCTCAAATTGCTA--------TTGTAGAATTAAA--ATCTATCATAGCAGCCGGTTTTTCATGTGTTATGCATGTTCATACAGCAATTGAAGAGGTACATATTGTTAAGTTATTGCACAAATTAGAAAAGGGTACC-------AACCGTAAGTCAAAGAAACCACCTGCTTTTGCTAAGAAGGGTATGAAGGTCATCGCTGTTTTAG--AAACTGAAGCTCCAGTTTGTGTGGAAACTTACCAA---GATTACCCTCAATTAGGTAGATTCACTTTGAGAGATCAAGGTACCACAATAGCAATTGGTAA-------AATTGTTAAAATTGCCGAG---TAA---\n",
      ">SUP35_Spar_A12_Liti_\n",
      "TTAT----------------------------------------------------------------------TCAGCAA------TTTTAACAATCTTACCAATTGCTATTGTGGTACCTTGATCTCTCAAAGTGAATCTACCTAA-----------TTGAGGGTA---ATCTTGGTAAGTTTCAACA---CAAACTGGAGCTTCAGTCTC----------TAAGACAGCAATGACCTTCATACCTTTCTTAGCAAACGCAGGTGGTTTCTTTGATTTACGGTTAGTACCCT----------TTTCCAATTTGTGCAATAATTTAACAATATGAACTTCTTCAATTGCTGTATGAACATGCATAACGCATGAAAACCCAGCAGCAATAATAGATTTTAA--------TTCGACAATTGCAATTTGAGCTACAAACTTGGTAACACTTTTGATAGGGTTCTTTGGTGATG------------------------------------TTAGTACGAACCCT------GGTGAAATGTCTTCTTCTTCAACACCTTTG-----------ATTCTTAGTTTAA----CTTGTTCACCACACATAGCCA--TGTCAACTTCATTTTCTGTTTCGTTGTAAATGTTTTGAATTTCCACTGCAGTTTTGTTAGGCATC--------------AATAGCGTGGATTGGCCCTTTTTGATATGAC----CAGATTCAATCTTACCTTCAACGATGGTACCTAAATCCTTCATTTTAGCGGCAAT---------------AGGCAACATGAAAGGAGCATTGATGTGAC----GGTCGACGTGGTTCATTGTGTCCAA---------------------------------ATATTCCAACAG-------------AGTTGGGCCAGTGTACCATGGGCATTCTTTTGGGTCTACGTGATCTTTCAGGTTTGCACCACTGTAACCAGATACTGGCATAAATACAACATCCGTCTTGATATTGTAGCCAATTGCTCTCAAGAAATTGCTGACATTACTCACA----CATTGGTCATAACGTTCCTTAGACCAGTTAACAGTTG--GGTCATCCATCTTATTTACGACAACAACCATCTTATTAACACCTTGGGTCT-TGGCTAATAGAGCGTGTTCACGTGTTTGACCACCTCTTTCA----------AAACCAGTTTCGTACTCACCCTTTCTGGCGGAAATGACCAAAACACCAACATCAGCTTGAGAAGCACCACCGATCATCTCGGAAACGTACATTTTATGACCAGGGGCATCCAATATGGTATAACGCCTTTTTTCAGTTTCGAAGTAGGCTTTACCAACTTCGATAGT-----TTTACCATCATTTCTTTCTTCTTTGTTGGTATCCATGACCCATGACAAGTACCAACCCTGTCTT----------CCCGCATCCTTGGCCTCTCTTTCATATTTCTCAATGGTTCTCTTATCGACAGAGCCAGT-------CAAGTACAGTAGATTACCACCCATAGTAGATT--TACCAGCATCAACATGACCCATGAAAATTAAGGAAACGTGATCTTTACCCCCGAACATGTCGTTAACAACTTCGTCATCGACTTCTTCTTCCTGTTCTTTGATCAAGGCATCAGCA-CTGTTAACATTGGCATTG--TTGGTGTTATCTGTTGATTCAGA-----------GATTTTCAGGTCTTCTACCTTTGGGAGTTCCGATT-----------TCTCCTCCTTTTTTTCTTCGGTCTTGACTAGTTTCTCCTCTTTTTTAACTGGTTCTTCGAGTTTTGTTGGTTCT-TTGTTTTCGGTGGGTTTGTCTTCCTTCTTCTTTTCAGATTCAGCAGGTT-------------TTGTGTC------GACCTTCTTGGTAGCATTGGCCAACTTGATACCAGAACTGGAAACAAGCTTCAAAGTTTTCTTTGGCTTGGGAGCGGCTTGCTTTTGTTGCTTTTGGAAGTCGTTCA-----AAGACATACCTTGAGACTGTGGTTGGAAACCAGCTTGATATCCTTGTGCATTGTTATTGTAGTTGAAGTTTTTGT---AATTGCCACGGCCACCTTGT-GGATTGAACTGTTGCTGATAACCGC------CTTGAGGATTGTAC-TGTTGATAGCCACCTTGGGGA-----------TTATATTGTTGC------TGGTAACCGGCATCTGGGTTGTACTGTTGGT------AGCTGCCT--TGTTGGTATCCAGAATA-----ACCTTGGTAATTTTGGTAATACCCACCTGCAGGTTGGGATTGAGCATTGTAAGCTTGATAACCTTGGTATCTGTTGTTACCTTGTTGTTGGTTAGGGTTTTGGCCGTATTGCTGGTAGCTTTGCTGATTGTTACCTTGGTTTGAATCCGACAT\n",
      ">SUP35_Smik_IFO1815T_30\n",
      "ATGTCTG---------------------------ATT--CAAA-------------------CCAAGGTAATAATCAGCAAAAC---TACCAGCAATACAACCAA----AACCCCAACCAACAG------CAAGGTAACAATAGATA------------CCAAGGTTATCAAGCTTATAATGCTCAAGCG---CAACCTGCAAGTGGCCATTAC---------CAAAACTACCAAGGCTACTCTGGGTACC--AACAAGGC---G-----CTT---TTCAACAGTACAACCCAG---AGGCTGGCTACCAA------CAACAGTACAAC-----------CCCCAAGGTGGCTATCAACAATACAATTCCCAGGGCGGTTATCAGCAACAATTCAATCCACAAGGCGGTCGTGGAAATTACCA---AAACATCAATTATAACAGCAATATACA-AGGATACCAAGCTGGTT------------------------------------TTCAACCACAGTCT------CAAAGTATGTCT-----TTAAATGACTTCCAAAA-------GCAACAAAAGCAA----GCTGCGCCTAAGCCAAAGAAGACTTTGAAGCTTGTTTCCAGTTCTGGTATCAAGCTAGCTAATGCTACCAAGAAAGTTGATACAAAACCTTCCGAAACTGAAAAGAAAGAGGA--AGTCAAGCCTGTCGAAAC--CAAAAAACCAATTAAA------------GTTGAAGAACCAGTCAAGAAAGAGGAAATGCCAGTAAAGGCCGAAGAAAAGAAGGAAGAAAAGTCTGAACTGCCAAAAGTAGGAG-ACTTAAAAATTTCTGA---------------------------------ATCGACAGATAACAGC---------AACACTACCAGTGCT---------------------------ACCAGTGCTGAT------GCCTTGATCAAAGA---ACAGGAAGAAGAAGTTGATGATGAGGTTGTTAATGATATGTTCGGGGGCAAAGATCACGTTTCTTTAATCTTTATGGGTCATGTCGATGCT--GGTAAGTCTACTATGGGTGGTAACTTACTGTACTT--GAC----------------------AGGCTCTGTGGATAAGAGAAC-TATAGAGAAATATGAAAGAGAGGCCAAGGATGCTGGTAGACAAGGTTGGTATTTGTCATGGGTCATGGATACCAACAAAGAAGAAAGA--AATGATGGTAAGACCATTG---AAGTTGGTAAGGCCTACTTCGAAACTGAGAAAAGGCGTTATACTATATTGGATGCTCCTGGTCATAAAATGTACGTTTCCGAAATGATTGGTGGTGCTTCTCAAGCTGATGTCGGTGTTTTAGTCATTTCTGCCAGAAAGGGTGAATACGAGACTGGTTTTGAAAGAGGTGGTCAAACTCGTGAACATGCCT---TATTGGCCAAGACCCAAGGTGTTAATAAGATGGTTGTCGTCGTAAATAAAATGGATGAT--CCAACTGTCAATTGGTCCAAGGAACGTTATGATCAATG----TGTAAGTAATGTCAGCAATTTCTTGAAAGCCATTGGCTATAACATCAAGACGGATGTTGTGTTTATG----------------CCAGTATCAGGTTACAGTGGTGCAAACTTGAAAAACCGTGTAGATCCAAAAGAATGTCCATGGTATACTGGCCCAACTCTATTGGAGTATTTAGATACAATGAACCATGTCGATCGTCACATCAACGCTCCATTCATGTTGCCTATTGCCGCTAA--GAT-----GAAGGATCTGGGCACGATCGTTGAAGGTAAGATTG--------AATCTGGTCATATCAAGAAGGGTCAATCCACCCTATTAATGCCCAACAAAACGGCAGTGGAAATTCAAAACA--TTTATAATGAAACTGAAAACGAAGTTGACA--TGGCTATGTGTGGTGAGCAAGTCAA----GCTGAGAATTAAGGGTGTTGAAGAAGAAGATATTTCCCCAGGGTTTGTACTAACATCACCAAAGAATCCT-----ATCAAGAGTGTCACCAAGTTTGTGGCCCAAATTGCCA--------TTGTCGAACTAAA--ATCTATTATAGCTGCTGGGTTTTCATGTGTTATGCATGTTCATACAGCAATTGAAGAAGTTCATATTGTTAAATTATTGCACAAATTGGAAAAGGGTACT-------AACCGTAAATCAAAGAAGCCACCTGCTTTTGCCAAGAAGGGTATGAAAGTCATTGCTGTTTTAG--AGACTGAAGCTCCGGTTTGTGTCGAAACTTACCAG---GATTACCCTCAGTTAGGTAGATTCACCTTAAGAGATCAAGGTGCTACAATAGCAATTGGTAA-------GATCGTTAAGATTGCTGAA---TGA---\n",
      ">SUP35_Skud_IFO1802T_36\n",
      "ATGTCAG---------------------------ATC--CAAA-------------------TCAAGGTAACAATCAACAA------------CAATACGGTCAA----AATCCTAACCAACAG------CAAGGCAATAACAAGTA------------CCAAGGTTATCAAGCTTACAATGCTCAAGCC---CAACCTGCAAACGGCTACTAT---------CAAAACTACCAAGGATTTGCCGGCTATC--AACAAGGC---G-----GCT---ACCAACAGTATGACCGAG---ACGCCGGTAGTCAG------CAACAATACAAC-----------CCACAAGGTGGCTATCAACAGTACCCCCCACAGGAAGGTTACCAACAACAGTCCAATCCACAAGGTGGTCGTGGTAATTACAA---AAGCTTCAACTATAACAACAATGTTCA-AGGGTATCAAGCTGGCT------------------------------------TCCAACCGCAATCT------CAAGGTATGTCT-----CTGAACGACTTCCAAAA-------GCAACAAAAGCAA----ACTGCTCCTAAGCCAAAGAAGACTTTGAAGCTTGTTTCAAGCTCTGGTATCAAGTTAGCTAATGCTACGAAGAAGGTCAGCACAAAGCCTGCCGAAACCGAAAAGAAAGAGGA--AGACAAGCCTGTCAAAAC--CGAGGAACAAATCAAG------------GTCGAAGAACCAATTAAAGAAGAGAAAAAAGCAGTCAAGACTGAGGAAAAGAAAGAGGTAGAATCTGAACTACCAAGGGTCGAAG-ACTTGAAGATATCCGA---------------------------------ATCGACTGATAATAGC---------AATACTGCTAACGTC---------------------------ACAAGTGCAGAC------ACTTTGATAAAAGA---ACAAGAAGAAGAAGTTGATGACGAAGTTGTTAATGATATGTTTGGCGGTAAAGATCACGTTTCTTTGATTTTCATGGGACATGTTGATGCT--GGTAAGTCTACTATGGGTGGTAATCTGCTATACTT--GAC----------------------TGGCTCCGTGGATAAAAGAAC-AATTGAAAAATACGAAAGAGAAGCTAAGGATGCTGGTAGGCAAGGTTGGTACTTGTCATGGGTCATGGATACAAATAAGGAAGAGAGA--AACGACGGTAAGACCATTG---AAGTTGGTAAAGCTTATTTCGAAACTGAGAAAAGACGTTATACCATTCTGGATGCTCCTGGTCATAAAATGTACGTTTCCGAAATGATTGGTGGTGCTTCGCAAGCCGATGTTGGTGTTTTAGTCATATCTGCCAGAAAGGGTGAATATGAAACCGGTTTCGAAAGAGGTGGCCAAACACGTGAACACGCCC---TATTGGCCAAGACCCAGGGTGTTAATAAAATGGTTGTTGTCGTAAATAAGATGGATGAT--GCAACCGTCAACTGGTCCAAAGAACGTTACGATCAATG----TGTAGGTAATGTTAGCAACTTTTTGAAAGCGATTGGTTACAACATCAAGACAGATGTTGCGTTTATG----------------CCAGTATCAGGTTACAGCGGTGCAAACTTGAAAAATCGTGTAGATCCAAAGGAGTGTTCATGGTATACCGGCCCAACTCTGTTGGAATACCTGGACACAATGAACCACGTCGATCGTCATGTCAATGCTCCATTCATGTTGCCTATTGCTGCTAA--GAT-----GAAGGATTTAGGTACCATAGTTGAAGGCAAGATCG--------AATCCGGTCATATCAAGAAGGGTCAATCCACCTTGTTGATGCCTAACAAAACCTCAGTGGAAATTCAAAACA--TTTACAACGAAACTGAAAACGAAGTTGACA--TGGCTATGTGCGGTGAACAGGTTAA----GTTGAGAATTAAAGGAGTTGAAGAAGAAGACATCTCTCCAGGGTTTGTGTTAACATCACCAAAGAACCCT-----ATCAAAAGTGTTACCAAATTTGTTGCCCAAATTGCCA--------TTGTAGAGTTAAA--GTCCATTATAGCTGCTGGGTTTTCATGTGTCATGCACGTTCATACAGCCATTGAAGAGGTTCATATTGTTAAATTATTGCACAAATTAGAAAAGGGTAGC-------AATCGTAAATCGAAGAAGCCACCTGCTTTCGCTAAGAAAGGTATGAAGGTTATTGCTGTTTTAG--AGACTGAAGCTCCAGTTTGTGTCGAAGCTTACCAA---GATTATCCTCAATTGGGTAGATTCACTCTAAGAGATCAAGGTACCACAATCGCCATCGGTAA-------AATTGTTAAGATTGCTGAA---TAA---\n",
      ">SUP35_Sbou_unique28_CM003560\n",
      "ATGTCGG---------------------------ATT--CAAA-------------------CCAAGGCAACAATCAGCAAAAC---TACCAGCAATACAGCCAG----AACGGTAACCAACAA------CAAGGTAACAACAGATA------------CCAAGGTTATCAAGCTTACAATGCTCAAGCC---CAACCTGCAGGTGGGTACTAC---------CAAAATTACCAAGGTTATTCTGGGTACC--AACAAGGT---G-----GCT---ATCAACAGTACAATCCCG---ACGCCGGTTACCAG------CAACAGTATAAT-----------CCTCAAGGAGGCTATCAACAGTACAATCCTCAAGGCGGTTATCAGCAGCAATTCAATCCACAAGGTGGCCGTGGAAATTACAA---AAACTTCAACTACAATAACAATTTGCA-AGGATATCAAGCTGGTT------------------------------------TCCAACCACAGTCT------CAAGGTATGTCT-----TTGAACGACTTTCAAAA-------GCAACAAAAGCAG----GCCGCTCCCAAACCAAAGAAGACTTTGAAGCTTGTCTCCAGTTCCGGTATCAAGTTGGCCAATGCTACCAAGAAGGTTGACACAAAACCTGCCGAATCTGATAAGAAAGAGGA--AGAGAAGTCTGCTGAAAC--CAAAGAACCAACTAAAGAGCCAACAAAGGTCGAAGAACCAGTTAAAAAGGAGGAGAAACCAGTCCAGACTGAAGAAAAGACGGAGGAAAAATCGGAACTTCCAAAGGTAGAAG-ACCTTAAAATCTCTGA---------------------------------ATCAACACATAATACC---------AACAATGCCAATGTT---------------------------ACCAGTGCTGAT------GCCTTGATCAAGGA---ACAGGAAGAAGAAGTGGATGACGAAGTTGTTAACGATATGTTTGGTGGTAAAGATCACGTTTCTTTAATTTTCATGGGTCATGTTGATGCC--GGTAAATCTACTATGGGTGGTAATCTACTATACTT--GAC----------------------TGGCTCTGTGGATAAGAGAAC-TATTGAGAAATATGAAAGAGAAGCCAAGGATGCAGGCAGACAAGGTTGGTACTTGTCATGGGTCATGGATACCAACAAAGAAGAAAGA--AATGATGGTAAGACTATCG---AAGTTGGTAAGGCCTACTTTGAAACTGAAAAAAGGCGTTATACCATATTGGATGCTCCTGGTCATAAAATGTACGTTTCCGAGATGATCGGTGGTGCTTCTCAAGCTGATGTTGGTGTTTTGGTCATTTCCGCCAGAAAGGGTGAGTACGAAACCGGTTTTGAGAGAGGTGGTCAAACTCGTGAACACGCCC---TATTGGCCAAGACCCAAGGTGTTAATAAGATGGTTGTCGTCGTAAATAAGATGGATGAC--CCAACCGTTAACTGGTCTAAGGAACGTTACGACCAATG----TGTGAGTAATGTCAGCAATTTCTTGAGAGCAATTGGTTACAACATTAAGACAGACGTTGTATTTATG----------------CCAGTATCCGGCTACAGTGGTGCAAATTTGAAAGATCACGTAGATCCAAAAGAATGCCCATGGTACACCGGCCCAACTCTGTTAGAATATCTGGATACAATGAACCACGTCGACCGTCACATCAATGCTCCATTCATGTTGCCTATTGCCGCTAA--GAT-----GAAGGATCTAGGTACCATCGTTGAAGGTAAAATTG--------AATCCGGTCATATCAAAAAGGGTCAATCCACCCTACTGATGCCTAACAAAACCGCTGTGGAAATTCAAAATA--TTTACAACGAAACTGAAAATGAAGTTGATA--TGGCTATGTGTGGTGAGCAAGTTAA----ACTAAGAATCAAAGGTGTTGAAGAAGAAGACATTTCACCAGGTTTTGTACTAACATCGCCAAAGAACCCT-----ATCAAGAGTGTTACCAAGTTTGTAGCTCAAATTGCTA--------TTGTAGAATTAAA--ATCTATCATAGCAGCCGGTTTTTCATGTGTTATGCATGTTCATACAGCAATTGAAGAGGTACATATTGTTAAGTTATTGCACAAATTAGAAAAGGGTACC-------AACCGTAAGTCAAAGAAACCACCTGCTTTTGCTAAGAAGGGTATGAAGGTCATCGCTGTTTTAG--AAACTGAAGCTCCAGTTTGTGTGGAAACTTACCAA---GATTACCCTCAATTAGGTAGATTCACTTTGAGAGATCAAGGTACCACAATAGCAATTGGTAA-------AATTGTTAAAATCGCCGAG---TAA---\n",
      ">SUP35_Scer_beer078_CM005938\n",
      "ATGTCGG---------------------------ATT--CAAA-------------------CCAAGGCAACAATCAGCAAAAC---TACCAGCAATACAGCCAG----AACGGTAACCAACAA------CAAGGTAACAACAGATA------------CCAAGGTTATCAAGCTTACAATGCTCAAGCC---CAACCTGCAGGTGGGTACTAC---------CAAAATTACCAAGGTTATTCTGGGTACC--AACAAGGT---G-----GCT---AT---------------------------------------CAACAGTACAAT-----------CCTCAA---------------------------GGCGGTTATCAGCAGCAATTCAATCCACAAGGTGGCCGTGGAAATTACAA---AAACTTCAACTACAATAACAATTTGCA-AGGATATCAAGCTGGTT------------------------------------TCCAACCACAGTCT------CAAGGTATGTCT-----TTGAACGACTTTCAAAA-------GCAACAAAAGCAG----GCCGCTCCCAAACCAAAGAAGACTTTGAAGCTTGTCTCCAGTTCCGGTATCAAGTTGGCCAATGCTACCAAGAAGGTTGACACAAAACCTGCCGAATCTGATAAGAAAGAGGA--AGAGAAGTCTGCTGAAAC--CAAAGAACCAACTAAAGAGCCAACAAAGGTCGAAGAACCAGTTAAAAAGGAGGAGAAACCAGTCCAGACTGAAGAAAAGACGGAGGAAAAATCGGAACTTCCAAAGGTAGAAG-ACCTTAAAATCTCTGA---------------------------------ATCAACACATAATACC---------AACAATGCCAATGTT---------------------------ACCAGTGCTGAT------GCCTTGATCAAGGA---ACAGGAAGAAGAAGTGGATGACGAAGTTGTTAACGATATGTTTGGTGGTAAAGATCACGTTTCTTTAATTTTCATGGGTCATGTTGATGCC--GGTAAATCTACTATGGGTGGTAATCTACTATACTT--GAC----------------------TGGCTCTGTGGATAAGAGAAC-TATTGAGAAATATGAAAGAGAAGCCAAGGATGCAGGCAGACAAGGTTGGTACTTGTCATGGGTCATGGATACCAACAAAGAAGAAAGA--AATGATGGTAAGACTATCG---AAGTTGGTAAGGCCTACTTTGAAACTGAAAAAAGGCGTTATACCATATTGGATGCTCCTGGTCATAAAATGTACGTTTCCGAGATGATCGGTGGTGCTTCTCAAGCTGATGTTGGTGTTTTGGTCATTTCCGCCAGAAAGGGTGAGTACGAAACCGGTTTTGAGAGAGGTGGTCAAACTCGTGAACACGCCC---TATTGGCCAAGACCCAAGGTGTTAATAAGATGGTTGTCGTCGTAAATAAGATGGATGAC--CCAACCGTTAACTGGTCTAAGGAACGTTACGACCAATG----TGTGAGTAATGTCAGCAATTTCTTGAGAGCAATTGGTTACAACATTAAGACAGACGTTGTATTTATG----------------CCAGTATCCGGCTACAGTGGTGCAAATTTGAAAGATCACGTAGATCCAAAAGAATGCCCATGGTACACCGGCCCAACTCTGTTAGAATATCTGGATACAATGAACCACGTCGACCGTCACATCAATGCTCCATTCATGTTGCCTATTGCCGCTAA--GAT-----GAAGGATCTAGGTACCATCGTTGAAGGTAAAATTG--------AATCCGGTCATATCAAAAAGGGTCAATCCACCCTACTGATGCCTAACAAAACCGCTGTGGAAATTCAAAATA--TTTACAACGAAACTGAAAATGAAGTTGATA--TGGCTATGTGTGGTGAGCAAGTTAA----ACTAAGAATCAAAGGTGTTGAAGAAGAAGACATTTCACCAGGTTTTGTACTAACATCGCCAAAGAACCCT-----ATCAAGAGTGTTACCAAGTTTGTAGCTCAAATTGCTA--------TTGTAGAATTAAA--ATCTATCATAGCAGCCGGTTTTTCATGTGTTATGCATGTTCATACAGCAATTGAAGAGGTACATATTGTTAAGTTATTGCACAAATTAGAAAAGGGTACC-------AACCGTAAGTCAAAGAAACCACCTGCTTTTGCTAAGAAGGGTATGAAGGTCATCGCTGTTTTAG--AAACTGAAGCTCCAGTTTGTGTGGAAACTTACCAA---GATTACCCTCAATTAGGTAGATTCACTTTGAGAGATCAAGGTACCACAATAGCAATTGGTAA-------AATTGTTAAAATTGCCGAG---TAA---\n",
      ">SUP35_Sarb_H-6_chrXIII_CM001575\n",
      "ATGTCTG---------------------------ATC--CAAC-------------------TAATGGTAATAATGAGCAGAGC---TCTCAACAGCAAGGCCAA----AACCCTAGCCAACAG------CAAGGTAACAACAGATA------------TCAAGGCTATCAAGCTTATAATGCTCAAACT---CAACCAGCAGGTGGCTATTAC---------CAAAACTACCAAGGCTATGCTGGCTACC--AACAAGGT---G-----GAT---ACCAACAGTTCACTCCAG---AGGCTGGTTACCAA------CAACAGTACAAC-----------CCCCAAGGTGGCTATCAACAGTTCAACCCCCAAGGCGGTTATCAACAACAGTTCAATCCACAAGGTGGCCGTGGCAACTACAA---AAACTTTAA---CAATAACAACCAACA-AGGATACCAAAGTAATT------------------------------------TTCAACCGCAATCT------CAAGGTATGTCT-----TTAAACGATTTCCAAAA-------GCAACAAAAGCAA----TCCACTCCTAAACCGAAGAAAACTTTGAAGCTTGTTTCAAGTTCCGGTATCAAGTTGGCTAATGCTACTAAGAAGGTCGATACAAAGCCCGTCGAAACCGAAAAGAAAGAAGA--AGACAAGCCTGTTGAAAC--TAAAAAACCAACGAAG------------GTCGAAGAACCAGCTAAGAAAGAGGAAGAGCCCGTCAAGGCTGAAGAAGCAAAGGAGGAAAAATCAGAACTACCAAAGGTTGAAG-ACTTAAAAATATCTGA---------------------------------ACCAACCGATAACAGC---------AACGCTGCTAGCGTC---------------------------AACAATGCAGAC------GCCTTAATTAAAGA---ACAAGAAGAAGAAGTCGATGATGAAGTTGTCAACGATATGTTTGGGGGCAAAGATCACGTTTCTTTGATCTTCATGGGTCATGTCGATGCT--GGTAAGTCCACGATGGGTGGTAATCTGCTTTACTT--GAC----------------------CGGGTCCGTGGATAAGAGAAC-TATTGAAAAATATGAAAGAGAAGCTAAGGATGCCGGTAGACAAGGTTGGTATTTGTCATGGGTCATGGATACCAACAAAGAAGAAAGA--AATGATGGTAAGACCATTG---AAGTTGGTAAAGCTTATTTCGAAACTGAAAAAAGACGTTATACCATTTTGGATGCTCCAGGTCATAAAATGTACGTTTCAGAAATGATTGGTGGTGCCTCTCAAGCTGATGTTGGTGTCTTAGTAATTTCTGCCAGAAAGGGTGAATATGAAACTGGTTTCGAAAGAGGTGGTCAAACACGTGAACACGCTT---TATTGGCCAAAACTCAAGGTGTTAATAAAATGGTTGTTGTCGTAAATAAGATGGACGAC--CCAACTGTCAACTGGTCCAAGGAACGTTACGACCAATG----TGTAAGTAATGTCAGCAATTTCTTAAAGGCCATTGGTTACAATATCAAGACTGATGTTGTATTCATG----------------CCAGTATCAGGTTACAGTGGTGCAAACTTGAAAGAACATGTAAATCCAAAAGAATGTTCATGGTACACCGGGCCAACTCTGTTGGAATACTTGGATAAAATGAATCACGTCGACCGTCACATCAATGCTCCATTCATGTTACCTATTGCTGCTAA--GAT-----GAAAGATCTAGGTACCATTGTAGAAGGTAAGATCG--------AATCCGGTCACATAAAGAAGGGTCAATCAACTCTTTTAATGCCTAACAAAACATCAGTGGAAATTCAAAATA--TTTACAATGAAACTGAAAACGAAGTTGATA--TGGCTATGTGTGGTGAACAAGTTAA----ATTAAGAATTAAAGGTGTTGAAGAAGAAGATATTTCACCAGGGTTCGTCTTGACATCACCAAAGAACCCA-----ATTAAGAGTGTTACCAAGTTTGTAGCTCAAATTGCCA--------TTGTTGAACTGAA--ATCTATTATAGCTGCTGGGTTTTCATGTGTTATGCACGTTCACACAGCCATTGAAGAAGTTCATATCGTTAAGCTATTGCACAAGTTAGAAAAGGGTACT-------AACCGTAAATCGAAGAAGCCACCTGCTTTTGCTAAGAAGGGTATGAAGGTCATTGCTGTTCTAG--AGACTGAAGCTCCAGTTTGTGTCGAGACTTACCAA---GACTACCCTCAATTGGGTAGATTCACTTTAAGAGACCAAGGTACCACAATCGCCATTGGTAA-------GATTGTTAAGATTGCTGAA---TAA---\n",
      ">SUP35_Seub_CBS12357_chr_II_IV_DF968535\n",
      "ATGTCTG---------------------------ATC--CAAA-------------------CCAAGGTAACAATCAGCAAAAC---TATCAACAGTACGGTCAA----AACTTCAACCAACAG------CAAGGTAACAACAAATT------------TCAAGGTTACCAAGCTTACAATGCTCAAGCCCAACAACCTGCAGGTGGCTATTAC---------CAAAACCCCCAAGGTTACGCTGGCTACC--AACAGGGC---G-----GTTATGACCAACAATTTAACCCGG---AAGTAGGTTACCAA------CAACAATACAAC-----------GCCCAA---------------------------GGTGGTTACCAACAACAGTTCAATCCACAAGGTGGCCGTGGCAATTACAA---GAACTTCAACTACAACAACAGCCAACA-GGGATTCCAAGCTGGCT------------------------------------TTCAACCACAATCT------CAAGGAATGTCT-----TTGAACGACTTCCAAAA-------GCAACAAAAACAA----ACTGCTCCTAAGCCAAAGAAGACTTTAAAACTTGTTTCAAGTTCCGGTATCAAGTTAGCTAATGCGACCAAGAAGGTCGACACAAAGCCTGTCGAAACTGAGAAGAAAGAGGA--AGAAAAGCCTACCGAAAC--CAAAGAACCAGCCAAG------------GTCGAAGAATCAATCAAAGATGTGGAAACTCCAGCCAGCGCTGAAGAAAAGAAGGAGGTGAATTCTGAATTACCAAAGGTCGAAG-ATTTGAAAATATCCGA---------------------------------ATCGAACGATAACAGC---------AAGCCTGCTAACCCT---------------------------ATCAACGCCGAT------GCCTTGATCAAAGA---ACAAGAAGATGAAGTAGACGATGAAGTTGTTAATGATATGTTCGGAGGTAAAGATCACGTTTCTTTAATTTTCATGGGTCACGTCGATGCT--GGTAAGTCCACTATGGGTGGTAATCTACTATATTT--GAC----------------------TGGCTCTGTGGACAAGAGAAC-CATTGAAAAATATGAAAGAGAAGCTAAAGATGCTGGTAGACAAGGTTGGTATTTGTCTTGGGTCATGGACACCAATAAAGAAGAAAGA--AACGACGGTAAGACCATTG---AGGTCGGTAAGGCTTATTTCGAAACCGAAAAAAGACGTTACACCATTCTGGATGCACCAGGTCATAAAATGTACGTTTCCGAAATGATTGGTGGTGCCTCTCAAGCCGATGTTGGTGTTCTAGTCATTTCTGCCAGAAAGGGTGAATACGAAACCGGTTTCGAAAGAGGTGGTCAAACACGTGAACACGCCT---TATTGGCCAAGACCCAAGGTGTTAATAAGATGATCGTTGTCGTAAATAAGATGGATGAT--CCTACTGTCAAATGGTCCAAAGAACGTTACGACCAATG----TGTAGGTAACGTCAGCAATTTCTTAAAGGCCATCGGTTACAATATAAAGACTGATGTTATATTCATG----------------CCAGTATCAGGTTACAGTGGTGCTAACTTGAAAGAACATGTAGATCCAAAGGAATGTTCATGGTACACCGGCCCAACTTTACTGGAATACTTGGACAAAATGACCCACGTTGACCGTCGTATCAACGCTCCATTCATGTTACCTATTGCTGCTAA--AAT-----GAAGGATTTAGGTACAATCGTAGAAGGTAAGATTG--------AATCTGGTCACATCAAGAAGGGCCAATCCACTTTATTGATGCCTAATAAGACCACTGTGGAAATCCAAAACA--TCTATAACGAAACTGAAACCGAAGTCGATA--TGGCCATGTGTGGTGAACAAGTCAA----ACTAAGAATTAAAGGTGTCGAAGAAGACGACATTTCACCAGGGTTTGTGTTGACATCCCCAAAGAACCCA-----ATTAAGAATGTTACCAAGTTTGTGGCGCAAATTGCCA--------TTGTCGAATTAAA--GTCCATTATAGCCGCCGGATTTTCATGTGTTATGCATGTTCACACAGCCATTGAAGAGGTTCATATTGTTAAACTACTGCATAAGTTAGAAAAGGGTACA-------AATCGTAAATCGAAGAAGCCACCTGCTTTTGCTAAGAAGGGTATGAAGGTCATTGCTGTTTTAG--AGACTGAAGCTCCAGTTTGCGTCGAGACTTACCAA---GATTATCCACAGTTGGGTAGATTTACACTTAGAGACCAAGGTACCACAATTGCCATTGGTAA-------GATTGTCAAGATTGCTGAA---TGA---\n"
     ]
    }
   ],
   "source": [
    "!cat SUP35_10seqs_strange_aln.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some long sequences here definetely ruin whole alignment. We could try to drop them or reverse or make them shorter (some of them align well at the beginning, others at the end)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. Run same aligners, but on longer sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# time_dct = {}\n",
    "# time_dct[\"clustalw\"] = %timeit -n1 -r1 -o !clustalw -INFILE=SUP35_250seqs.fa -OUTPUT=FASTA -OUTFILE=05_SUP35_250seqs_clustalw.fa > logs/logs_clustalw.txt\n",
    "# time_dct[\"muscle\"] = %timeit -n1 -r1 -o !muscle -align SUP35_250seqs.fa -output 05_SUP35_250seqs_muscle.fa 2> logs/logs_muscle_std2.txt 1> logs/logs_muscle_std1.txt\n",
    "# time_dct[\"mafft\"] = %timeit -n1 -r1 -o !mafft --auto SUP35_250seqs.fa > 05_SUP35_250seqs_mafft.fa 1> logs/logs_mafft_std1.txt 2> logs/logs_mafft_std2.txt\n",
    "# time_dct[\"kalign\"] = %timeit -n1 -r1 -o !kalign <SUP35_250seqs.fa> 05_SUP35_250seqs_kalign.fa 2> logs/logs_kalign.txt\n",
    "# time_dct[\"tcoffee\"] = %timeit -n1 -r1 -o !t_coffee -infile=SUP35_250seqs.fa -outfile=05_SUP35_250seqs_tcoffee.fa -output=fasta_aln 2> logs/logs_t_coffee_std2.txt 1> logs/logs_t_coffee_std1.txt\n",
    "# time_dct[\"prank\"] = %timeit -n1 -r1 -o !prank -d=SUP35_250seqs.fa -o=05_SUP35_250seqs_prank.fa > logs/logs_prank.txt\n",
    "\n",
    "# for k in time_dct:\n",
    "#     time_dct[k] = round(time_dct[k].average, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "time_dct = {\n",
    "    \"clustalw\": 1361.64,\n",
    "    \"muscle\": \"too long :(\", # option -super5 can be used, but i didnt try it\n",
    "    \"mafft\": 20.96,\n",
    "    \"kalign\": 12.63,\n",
    "    \"tcoffee\": 26261.42,\n",
    "    \"prank\": 185.78,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6. Comparative table (time and aln length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "length_dct = {}\n",
    "\n",
    "for file in os.listdir():\n",
    "    if file.startswith(\"05_SUP35_250seqs\"):\n",
    "        length_dct[file.split(\"_\")[-1].split(\".\")[0]] = get_aln_len(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mafft': 2341,\n",
       " 'tcoffee': 2269,\n",
       " 'clustalw': 2207,\n",
       " 'prank': 2480,\n",
       " 'kalign': 2233,\n",
       " 'muscle': 0}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "length_dct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>time</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>kalign</th>\n",
       "      <td>12.63</td>\n",
       "      <td>2233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mafft</th>\n",
       "      <td>20.96</td>\n",
       "      <td>2341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prank</th>\n",
       "      <td>185.78</td>\n",
       "      <td>2480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clustalw</th>\n",
       "      <td>1361.64</td>\n",
       "      <td>2207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tcoffee</th>\n",
       "      <td>26261.42</td>\n",
       "      <td>2269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>muscle</th>\n",
       "      <td>too long :(</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 time  length\n",
       "kalign          12.63    2233\n",
       "mafft           20.96    2341\n",
       "prank          185.78    2480\n",
       "clustalw      1361.64    2207\n",
       "tcoffee      26261.42    2269\n",
       "muscle    too long :(       0"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame({\"time\": time_dct, \"length\": length_dct}).sort_values(\n",
    "    by=[\"time\", \"length\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Shorter alignment with best time was obtained with kalign again. But we might want to consider biological sense too."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "7. Getting aminoacid sequences from nucleotide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Translate nucleic acid sequences\n"
     ]
    }
   ],
   "source": [
    "!transeq -sequence SUP35_10seqs.fa -outseq SUP35_10seqs.t.faa\n",
    "# !cat SUP35_10seqs.t.faa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Find and extract open reading frames (ORFs)\n"
     ]
    }
   ],
   "source": [
    "!getorf -sequence SUP35_10seqs.fa -outseq SUP35_10seqs.g.faa -noreverse -minsize 500\n",
    "# !cat SUP35_10seqs.g.faa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We need to consider ORFs. Automatic ORF finders might be too \"mechanical\", if every start-codon is ORF opening for them."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "8. Aligning protein sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      " CLUSTAL 2.1 Multiple Sequence Alignments\n",
      "\n",
      "\n",
      "Sequence type explicitly set to Protein\n",
      "Sequence format is Pearson\n",
      "Sequence 1: SUP35_Kla_AB039749_1                       700 aa\n",
      "Sequence 2: SUP35_Agos_ATCC_10895_NM_211584_1          691 aa\n",
      "Sequence 3: SUP35_Scer_74-D694_GCA_001578265.1_1       685 aa\n",
      "Sequence 4: SUP35_Spar_A12_Liti_1                      681 aa\n",
      "Sequence 5: SUP35_Smik_IFO1815T_30_1                   681 aa\n",
      "Sequence 6: SUP35_Skud_IFO1802T_36_1                   678 aa\n",
      "Sequence 7: SUP35_Sbou_unique28_CM003560_1             685 aa\n",
      "Sequence 8: SUP35_Scer_beer078_CM005938_1              666 aa\n",
      "Sequence 9: SUP35_Sarb_H-6_chrXIII_CM001575_1          680 aa\n",
      "Sequence 10: SUP35_Seub_CBS12357_chr_II_IV_DF968535_1   674 aa\n",
      "Start of Pairwise alignments\n",
      "Aligning...\n",
      "\n",
      "Sequences (1:2) Aligned. Score:  71\n",
      "Sequences (1:3) Aligned. Score:  69\n",
      "Sequences (1:4) Aligned. Score:  69\n",
      "Sequences (1:5) Aligned. Score:  68\n",
      "Sequences (1:6) Aligned. Score:  68\n",
      "Sequences (1:7) Aligned. Score:  69\n",
      "Sequences (1:8) Aligned. Score:  70\n",
      "Sequences (1:9) Aligned. Score:  67\n",
      "Sequences (1:10) Aligned. Score:  69\n",
      "Sequences (2:3) Aligned. Score:  68\n",
      "Sequences (2:4) Aligned. Score:  69\n",
      "Sequences (2:5) Aligned. Score:  70\n",
      "Sequences (2:6) Aligned. Score:  69\n",
      "Sequences (2:7) Aligned. Score:  69\n",
      "Sequences (2:8) Aligned. Score:  71\n",
      "Sequences (2:9) Aligned. Score:  67\n",
      "Sequences (2:10) Aligned. Score:  69\n",
      "Sequences (3:4) Aligned. Score:  97\n",
      "Sequences (3:5) Aligned. Score:  94\n",
      "Sequences (3:6) Aligned. Score:  92\n",
      "Sequences (3:7) Aligned. Score:  99\n",
      "Sequences (3:8) Aligned. Score:  99\n",
      "Sequences (3:9) Aligned. Score:  92\n",
      "Sequences (3:10) Aligned. Score:  91\n",
      "Sequences (4:5) Aligned. Score:  94\n",
      "Sequences (4:6) Aligned. Score:  92\n",
      "Sequences (4:7) Aligned. Score:  97\n",
      "Sequences (4:8) Aligned. Score:  96\n",
      "Sequences (4:9) Aligned. Score:  93\n",
      "Sequences (4:10) Aligned. Score:  91\n",
      "Sequences (5:6) Aligned. Score:  92\n",
      "Sequences (5:7) Aligned. Score:  94\n",
      "Sequences (5:8) Aligned. Score:  94\n",
      "Sequences (5:9) Aligned. Score:  92\n",
      "Sequences (5:10) Aligned. Score:  90\n",
      "Sequences (6:7) Aligned. Score:  92\n",
      "Sequences (6:8) Aligned. Score:  91\n",
      "Sequences (6:9) Aligned. Score:  91\n",
      "Sequences (6:10) Aligned. Score:  90\n",
      "Sequences (7:8) Aligned. Score:  100\n",
      "Sequences (7:9) Aligned. Score:  93\n",
      "Sequences (7:10) Aligned. Score:  91\n",
      "Sequences (8:9) Aligned. Score:  92\n",
      "Sequences (8:10) Aligned. Score:  91\n",
      "Sequences (9:10) Aligned. Score:  91\n",
      "Guide tree file created:   [SUP35_10seqs.g.dnd]\n",
      "\n",
      "There are 9 groups\n",
      "Start of Multiple Alignment\n",
      "\n",
      "Aligning...\n",
      "Group 1: Sequences:   2      Score:14421\n",
      "Group 2: Sequences:   3      Score:14556\n",
      "Group 3: Sequences:   4      Score:14367\n",
      "Group 4: Sequences:   5      Score:14254\n",
      "Group 5: Sequences:   6      Score:14159\n",
      "Group 6: Sequences:   7      Score:14064\n",
      "Group 7: Sequences:   8      Score:13830\n",
      "Group 8: Sequences:   2      Score:12740\n",
      "Group 9: Sequences:  10      Score:11985\n",
      "Alignment Score 155481\n",
      "firstres = 1 lastres = 719\n",
      "FASTA file created!\n",
      "\n",
      "Fasta-Alignment file created    [08_SUP35_10seqs_clustalw.faa]\n",
      "\n",
      "413 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "FATAL: Cowardly refusing to overwrite already existing file '08_SUP35_10seqs_clustalo.faa'. Use --force to force overwriting.\n",
      "121 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "\n",
      "muscle 5.1.linux64 []  16.1Gb RAM, 8 cores\n",
      "Built May 16 2023 07:53:40\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 10 seqs, avg length 682, max 700\n",
      "\n",
      "00:00 7.9Mb  CPU has 8 cores, running 8 threads\n",
      "00:01 558Mb   100.0% Calc posteriors\n",
      "00:01 540Mb   100.0% Consistency (1/2)\n",
      "00:01 540Mb   100.0% Consistency (2/2)\n",
      "00:01 540Mb   100.0% UPGMA5           \n",
      "00:01 541Mb   100.0% Refining\n",
      "751 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.520\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "rescale = 1\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 10\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     8 /9 \n",
      "Reallocating..done. *alloclen = 2413\n",
      "STEP     9 /9 \n",
      "done.\n",
      "tbfast (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "rescale = 1\n",
      "\n",
      "    0 / 10\n",
      "Segment   1/  1    1- 753\n",
      "STEP 003-003-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (aa) Version 7.520\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "471 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "\n",
      "Kalign version 2.04, Copyright (C) 2004, 2005, 2006 Timo Lassmann\n",
      "\n",
      "        Kalign is free software. You can redistribute it and/or modify\n",
      "        it under the terms of the GNU General Public License as\n",
      "        published by the Free Software Foundation.\n",
      "\n",
      "reading from STDIN: found 10 sequences\n",
      "Aligning 10 protein sequences with these parameters:\n",
      "        54.94940948\tgap open penalty\n",
      "        8.52492046\tgap extension\n",
      "        4.42409992\tterminal gap penalty\n",
      "        0.20000000\tbonus\n",
      "Alignment will be written to stdout.\n",
      "\n",
      "Distance Calculation:\n",
      "     100 percent done\n",
      "Alignment:\n",
      "     100 percent done\n",
      "141 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n",
      "\n",
      "-----------------\n",
      " PRANK v.170427:\n",
      "-----------------\n",
      "\n",
      "Input for the analysis\n",
      " - aligning sequences in 'SUP35_10seqs.g.faa'\n",
      " - using inferred alignment guide tree\n",
      " - option '+F' is not used; it can be enabled with '+F'\n",
      " - external tools available:\n",
      "    MAFFT for initial alignment\n",
      "    Exonerate for alignment anchoring\n",
      "\n",
      "Warning: sequence names changed.\n",
      "\n",
      "Correcting (arbitrarily) for multifurcating nodes.\n",
      "Correcting (arbitrarily) for multifurcating nodes.\n",
      "\n",
      "Generating multiple alignment: iteration 1.\n",
      "\n",
      "Alignment score: 1285\n",
      "Correcting (arbitrarily) for multifurcating nodes.\n",
      "\n",
      "Generating multiple alignment: iteration 2.\n",
      "\n",
      "Alignment score: 1297\n",
      "Correcting (arbitrarily) for multifurcating nodes.\n",
      "\n",
      "Generating multiple alignment: iteration 3.\n",
      "\n",
      "Alignment score: 1292\n",
      "Correcting (arbitrarily) for multifurcating nodes.\n",
      "\n",
      "Generating multiple alignment: iteration 4.\n",
      "\n",
      "Alignment score: 1292\n",
      "Correcting (arbitrarily) for multifurcating nodes.\n",
      "\n",
      "Generating multiple alignment: iteration 5.\n",
      "\n",
      "Alignment score: 1292\n",
      "\n",
      "\n",
      "Writing\n",
      " - alignment to '08_SUP35_10seqs_prank.faa.best.fas'\n",
      "\n",
      "Analysis done. Total time 7s\n",
      "\n",
      "7.53 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "time_dct = {}\n",
    "time_dct[\"clustalw\"] = %timeit -n1 -r1 -o !clustalw -INFILE=SUP35_10seqs.g.faa -OUTFILE=08_SUP35_10seqs_clustalw.faa -OUTPUT=FASTA -TYPE=protein   \n",
    "time_dct[\"clustalo\"] = %timeit -n1 -r1 -o !clustalo --infile=SUP35_10seqs.g.faa --outfile=08_SUP35_10seqs_clustalo.faa # [--verbose]   \n",
    "time_dct[\"muscle\"] = %timeit -n1 -r1 -o !muscle -align SUP35_10seqs.g.faa -output 08_SUP35_10seqs_muscle.faa  \n",
    "time_dct[\"mafft\"] = %timeit -n1 -r1 -o !mafft --auto SUP35_10seqs.g.faa >08_SUP35_10seqs_mafft.fa  \n",
    "time_dct[\"kalign\"] = %timeit -n1 -r1 -o !kalign <SUP35_10seqs.g.faa> 08_SUP35_10seqs_kalign.faa  \n",
    "# time_dct[\"tcoffee\"] = %timeit -n1 -r1 -o !export MAX_N_PID_4_TCOFFEE=400000 && echo $MAX_N_PID_4_TCOFFEE && t_coffee -infile=SUP35_10seqs.g.faa -outfile=08_SUP35_10seqs_tcoffee.faa -output=fasta_aln  \n",
    "# couldn't manage to fix it :(\n",
    "time_dct[\"prank\"] = %timeit -n1 -r1 -o !prank -d=SUP35_10seqs.g.faa -o=08_SUP35_10seqs_prank.faa  \n",
    "\n",
    "for k in time_dct:\n",
    "    time_dct[k] = round(time_dct[k].average, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'clustalw': 0.41,\n",
       " 'clustalo': 0.12,\n",
       " 'muscle': 0.75,\n",
       " 'mafft': 0.47,\n",
       " 'kalign': 0.14,\n",
       " 'prank': 7.53}"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "time_dct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "9. Comparative table for protein aln"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "length_dct = {}\n",
    "\n",
    "for file in os.listdir():\n",
    "    if file.startswith(\"08_SUP35_10seqs\"):\n",
    "        length_dct[file.split(\"_\")[-1].split(\".\")[0]] = get_aln_len(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'clustalo': 789,\n",
       " 'kalign': 752,\n",
       " 'muscle': 807,\n",
       " 'prank': 823,\n",
       " 'mafft': 791,\n",
       " 'clustalw': 756}"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "length_dct"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "9. Comparative table for proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>time</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>clustalo</th>\n",
       "      <td>0.12</td>\n",
       "      <td>789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kalign</th>\n",
       "      <td>0.14</td>\n",
       "      <td>752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>clustalw</th>\n",
       "      <td>0.41</td>\n",
       "      <td>756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mafft</th>\n",
       "      <td>0.47</td>\n",
       "      <td>791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>muscle</th>\n",
       "      <td>0.75</td>\n",
       "      <td>807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prank</th>\n",
       "      <td>7.53</td>\n",
       "      <td>823</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          time  length\n",
       "clustalo  0.12     789\n",
       "kalign    0.14     752\n",
       "clustalw  0.41     756\n",
       "mafft     0.47     791\n",
       "muscle    0.75     807\n",
       "prank     7.53     823"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame({\"time\": time_dct, \"length\": length_dct}).sort_values(\n",
    "    by=[\"time\", \"length\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "10. Add new aligned sequences to existing aln with mafft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "All-to-all alignment.\n",
      "tbfast-pair (nuc) Version 7.520\n",
      "alg=L, model=DNA200 (2), 2.00 (6.00), -0.10 (-0.30), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 2\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     1 /1 \n",
      "done.\n",
      "tbfast (nuc) Version 7.520\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "\n",
      "    0 / 2\n",
      "Segment   1/  1    1-2059\n",
      "done 001-001-1  identical.   \n",
      "dvtditr (nuc) Version 7.520\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n",
      "nadd = 2\n",
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 12 ambiguous characters.\n",
      "  201 / 252\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "  250 / 252\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP   208 / 251 \n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "  200 / 252\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "  250 / 252\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP   209 / 251 \n",
      "done.\n",
      "\n",
      "disttbfast (nuc) Version 7.520\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-2 (Fast but rough)\n",
      " Progressive method (guide trees were built 2 times.)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#!muscle -align SUP35_2addseqs.fa -output 10_SUP35_2addseqs_muscle.fa\n",
    "# !muscle -profile -in1 05_SUP35_250seqs_muscle.fa -in2 SUP35_2addseqs.fa -out 10_SUP35_252seqs_muscle.fa\n",
    "# unfortunately muscle v5 doesnt seem to have this option anymore (or i couldnt find it)\n",
    "!mafft --auto SUP35_2addseqs.fa > 10_SUP35_2addseqs_mafft.fa\n",
    "!mafft --add 10_SUP35_2addseqs_mafft.fa --reorder 05_SUP35_250seqs_mafft.fa > 10_SUP35_252seqs_mafft.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "11. Extracting Parapallasea 18S sequences and saving to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "!esearch -db nucleotide -query \"Parapallasea 18S\" | efetch -format fasta > 11_Parapallasea_18.fa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "All-to-all alignment.\n",
      "tbfast-pair (nuc) Version 7.520\n",
      "alg=L, model=DNA200 (2), 2.00 (6.00), -0.10 (-0.30), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "    0 / 4\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP     3 /3 \n",
      "done.\n",
      "tbfast (nuc) Version 7.520\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "1 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "\n",
      "    0 / 4\n",
      "Segment   1/  1    1-2270\n",
      "STEP 002-002-1  identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "done\n",
      "dvtditr (nuc) Version 7.520\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!mafft --auto 11_Parapallasea_18.fa > 11_Parapallasea_18_mafft.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The resulting aligment has 3  well-aligned sequences, another aligns to the end. I suppose it's OK for rRNA genes to be that clustered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "12. Making custom db for blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 03/08/2024 16:27:13\n",
      "New DB name:   /home/alex/dev/phylo/hw_3/Ommatogammarus_flavus_transcriptome_assembly.fa\n",
      "New DB title:  Ommatogammarus_flavus_transcriptome_assembly.fa\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 3000000000B\n",
      "Adding sequences from FASTA; added 100 sequences in 0.00353789 seconds.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!makeblastdb -in Ommatogammarus_flavus_transcriptome_assembly.fa -dbtype nucl -parse_seqids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Acanthogammarus_victorii_COI\n",
      "MIKHWLFSTNHKDIGTLYFILGAWAGMVGTSMSVIIRSELSAPGNLIGDDQLYNVMVTSHAFVMIFFMVMPIMIGGFGNWLVPLMLGSPDMAFPRMNNMSFWLLPPSLTLLITSGLVESGVGTGWTVYPPLSGTTAHSGSAVDLAIFSLHLAGASSILGAVNFISTIANMRSPSMALDQMPLFVWAILITTVLLLLSLPVLAGAITMLLTDRNLNTSFFDPGGGGDPILYQHLFWFFGHPEVYILILPAFGMVSHIVSQESGKKETFGSLGMIYAMLAIGFLGFIVWAHHMFTVGMDVDTRAYFTSATMIIAVPTGIKVFSWLGTLQGGSIKFTPALIWSLGFIFLFSIGGLTGVMLANSSIDIVLHDTYYVVAHFHYVLSMGAVFGIFAGFAYWFPLFTGVTANPVLAKAHFYAMFTGVNLTFFPQHFLGLTGMPRRYSDYPDFFTAWNTMSSVGSYISVVGMVLFVLMVMEAFVAKRPALFSLSLASALEWQHPYPPAGHSYSSTPILAS"
     ]
    }
   ],
   "source": [
    "!cat Acanthogammarus_victorii_COI.faa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For mitochondria has sligtly different genecode, we need to specify it via `-db_gencode 5` option"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "!tblastn -query Acanthogammarus_victorii_COI.faa -db Ommatogammarus_flavus_transcriptome_assembly.fa -outfmt 6 -db_gencode 5 > Acanthogammarus_victorii_COI_blast_output_6.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acanthogammarus_victorii_COI</td>\n",
       "      <td>TRINITY_DN8878_c0_g1_i2</td>\n",
       "      <td>89.621</td>\n",
       "      <td>501</td>\n",
       "      <td>52</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>509</td>\n",
       "      <td>3</td>\n",
       "      <td>1505</td>\n",
       "      <td>0.0</td>\n",
       "      <td>781.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acanthogammarus_victorii_COI</td>\n",
       "      <td>TRINITY_DN58613_c0_g1_i1</td>\n",
       "      <td>50.000</td>\n",
       "      <td>20</td>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "      <td>206</td>\n",
       "      <td>225</td>\n",
       "      <td>32</td>\n",
       "      <td>88</td>\n",
       "      <td>6.3</td>\n",
       "      <td>20.8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         qseqid                    sseqid  pident  length  \\\n",
       "0  Acanthogammarus_victorii_COI   TRINITY_DN8878_c0_g1_i2  89.621     501   \n",
       "1  Acanthogammarus_victorii_COI  TRINITY_DN58613_c0_g1_i1  50.000      20   \n",
       "\n",
       "   mismatch  gapopen  qstart  qend  sstart  send  evalue  bitscore  \n",
       "0        52        0       9   509       3  1505     0.0     781.0  \n",
       "1         9        1     206   225      32    88     6.3      20.8  "
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\n",
    "    \"Acanthogammarus_victorii_COI_blast_output_6.txt\",\n",
    "    sep=\"\\t\",\n",
    "    header=None,\n",
    "    names=[\n",
    "        \"qseqid\",\n",
    "        \"sseqid\",\n",
    "        \"pident\",\n",
    "        \"length\",\n",
    "        \"mismatch\",\n",
    "        \"gapopen\",\n",
    "        \"qstart\",\n",
    "        \"qend\",\n",
    "        \"sstart\",\n",
    "        \"send\",\n",
    "        \"evalue\",\n",
    "        \"bitscore\",\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Best match to separate file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "!blastdbcmd -db Ommatogammarus_flavus_transcriptome_assembly.fa -entry TRINITY_DN8878_c0_g1_i2 -out Ommatogammarus_flavus_COI.fa "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">TRINITY_DN8878_c0_g1_i2 len=1505\n",
      "CGACCAACCACAAAGATATTGGCACTCTTTATTTTATGCTAGGGCTCTGGTCTGGGTTAGTCGGAACCTCCATAAGACTT\n",
      "ATCCTCCGCTCAGAACTTAGTGCGCCGGGTAGCCTGATTGGTGATGATCAACTGTATAACGTAATGGTAACCTCCCATGC\n",
      "TTTTATTATAATTTTTTTTATAGTTATGCCTATCATAATTGGCGGGTTTGGTAACTGGCTGCTTCCTTTAATACTAGGTA\n",
      "GACCTGATATAGCCTTCCCTCGAATAAACAACATGAGCTTTTGACTACTACCTCCTTCCCTTACACTTCTTATATCTAGA\n",
      "AGCTTAGTAGAAAGAGGAGTCGGCACAGGTTGAACTGTCTACCCTCCTTTATCTGGGTCTACAGCCCATAGAGGTAGCGC\n",
      "TGTAGATTTGGCTATTTTCTCACTTCATTTAGCCGGAGCTTCCTCTATCTTAGGGGCTGTAAATTTTATTTCTACCGCCA\n",
      "TTAATATGCGAGCGCCTGGGATAAAATTAGACCAAATGCCTTTATTCGTCTGAGCTATTATTATTACTACCGTCCTCCTA\n",
      "GTCTTATCCCTACCAGTCCTAGCTGGGGCCATTACGATACTACTTACAGACCGTAACATAAATACCTCTTTTTTTGACCC\n",
      "TAGTGGGGGGGGTGACCCTATCCTATACCAACACTTATTTTGATTTTTTGGGCACCCAGAGGTGTATATTTTAATCCTGC\n",
      "CTGCATTTGGCATAATCTCTCATATTGTTAGACAGGAGTCCGGTAAAAAAGAAACATTTGGCCCCCTAGGGATAATTTAT\n",
      "GCTATATTAGCTATTGGGTTCCTCGGATTTATTGTGTGAGCCCATCATATGTTTACAGTCGGTATGGATGTAGATACCCG\n",
      "AGCCTATTTTACATCAGCTACAATAATTATTGCAGTCCCCACCGGCATCAAAGTATTTAGGTGACTAGGTACTCTACAAG\n",
      "GCGGAAAAATTAACTTTTCTCCAGCTCTAATTTGAAGACTAGGTTTTATTTTCCTTTTCTCTATTGGAGGTTTAACTGGA\n",
      "GTTATATTAGCTAACTCATCAATTGACATCGTACTACACGACACTTACTATGTAGTTGCCCACTTTCATTATGTTTTATC\n",
      "TATGGGGGCTGTTTTCGGTATTTTTGCGGGGTTTGCTTACTGGTTTCCACTATTTACAGGTATAACTATCAATCCTATCC\n",
      "TAGCTAAAATTCATTTTTACGTCATATTCATGGGAGTAAACTTAACTTTTTTCCCCCAACATTTCCTTGGTTTAACGGGC\n",
      "ATACCTCGGCGATACTCAGACTATCCTGACTTCTTCACAGCCTGAAATATTGTTTCCTCCTTAGGCTCTTATATCTCTGT\n",
      "TTTAGCTATAGTGATCTTTATTGCTATAATCATAGAAGCTTTTATCTCTAAGCGGTCCGCTTTATTTTCCTTAACCTTGT\n",
      "CGTCTGCTTTAGAGTGGTACCACTCATACCCGCCAGCCGACCATAGCTACAACGATACCCCTATT\n"
     ]
    }
   ],
   "source": [
    "!cat Ommatogammarus_flavus_COI.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "phylo",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
